staging/src/k8s.io/apiserver/pkg/endpoints/discovery/v2/handler.go

```diff
@@ -0,0 +1,294 @@
+		rdm.apiGroups[groupName] = &metav1.APIGroupDiscovery{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "APIGroupDiscovery",
+				APIVersion: "v1",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: groupName,
+			},
+			Versions: []metav1.APIVersionDiscovery{value},
+		}
+		rdm.apiGroupNames = append(rdm.apiGroupNames, groupName)
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroupVersion(apiGroup metav1.GroupVersion) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	group, exists := rdm.apiGroups[apiGroup.Group]
+	if !exists {
+		return
+	}
+
+	for i := range group.Versions {
+		if group.Versions[i].Version == apiGroup.Version {
+			group.Versions = append(group.Versions[:i], group.Versions[i+1:]...)
+			break
+		}
+	}
+
+	if len(group.Versions) == 0 {
+		delete(rdm.apiGroups, group.Name)
+		for i := range rdm.apiGroupNames {
+			if rdm.apiGroupNames[i] == group.Name {
+				rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+				break
+			}
+		}
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) RemoveGroup(groupName string) {
+	rdm.lock.Lock()
+	defer rdm.lock.Unlock()
+
+	delete(rdm.apiGroups, groupName)
+	for i := range rdm.apiGroupNames {
+		if rdm.apiGroupNames[i] == groupName {
+			rdm.apiGroupNames = append(rdm.apiGroupNames[:i], rdm.apiGroupNames[i+1:]...)
+			break
+		}
+	}
+
+	// Reset response document so it is recreated lazily
+	rdm.cachedGroupList = cachedGroupList{}
+}
+
+func (rdm *resourceDiscoveryManager) WebService() *restful.WebService {
+	mediaTypes, _ := negotiation.MediaTypesForSerializer(rdm.serializer)
+	ws := new(restful.WebService)
+	ws.Path(DiscoveryEndpointRoot)
+	ws.Doc("get available API groupversions and resources")
+
+	ws.Route(ws.GET("/v2").To(func(req *restful.Request, resp *restful.Response) {
+		rdm.ServeHTTP(resp.ResponseWriter, req.Request)
+	}).
+		Doc("get available API groupversions and their resources").
+		Operation("getDiscoveryResources").
+		Produces(mediaTypes...).
+		Consumes(mediaTypes...).
+		Writes(metav1.APIGroupDiscoveryList{}))
+	return ws
+}
+
+func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
+	rdm.lock.RLock()
+	response := rdm.cachedResponse
+	etag := rdm.cachedResponseETag
+
+	// The cachedResponse is wiped out every time there might be a change to its
+	// contents.
+	if response != nil {
+		defer rdm.lock.RUnlock()
+	} else {
+		rdm.lock.RUnlock()
+
+		// Document does not exist, recreate it
+		rdm.lock.Lock()
+		defer rdm.lock.Unlock()
+
+		// Now that we have taken exclusive lock, check to see if another thread
+		// recreated the document while we were waiting for the lock
+		response, etag = rdm.cachedResponse, rdm.cachedResponseETag
+		if response == nil {
+			// Re-order the apiGroups by their insertion order
```

Comment:
Is a hold over from our previous means of prioritizing groups?  I thought we hardcoded a list somewhere to determine priority.

Reply:
right. that list ends up not getting used anyway due to aggregator. ive removed it.